include/ginkgo/core/base/batch_lin_op.hpp

```diff
@@ -0,0 +1,439 @@
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_BASE_BATCH_LIN_OP_HPP_
+#define GKO_PUBLIC_CORE_BASE_BATCH_LIN_OP_HPP_
+
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+
+#include <ginkgo/core/base/abstract_factory.hpp>
+#include <ginkgo/core/base/batch_lin_op_helpers.hpp>
+#include <ginkgo/core/base/dim.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/matrix_assembly_data.hpp>
+#include <ginkgo/core/base/matrix_data.hpp>
+#include <ginkgo/core/base/polymorphic_object.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/log/logger.hpp>
+
+
+namespace gko {
+namespace batch {
+
+
+/**
+ * @addtogroup BatchLinOp
+ *
+ * @section batch_linop_concept Batched Linear operator as a concept
+ *
+ * A batch linear operator (BatchLinOp) forms the base class for all batched
+ * linear algebra objects. In general, it follows the same structure as the
+ * LinOp class, but has some crucial differences which make it not strictly
+ * representable through or with the LinOp class.
+ *
+ * A batched operator is defined as a set of independent linear operators which
+ * have no communication/information exchange between them. Therefore, any
+ * collective operations between the batches is not possible and not
+ * implemented. This allows for each batch to be computed and operated on in an
+ * embarrasingly parallel fashion.
+ *
+ * Similar to the LinOp class, the BatchLinOp also implements
+ * BatchLinOp::apply() methods which call the internal apply_impl() methods
+ * which the concrete BatchLinOp's have to implement.
+ *
+ * A key difference between the LinOp and the BatchLinOp classes is the storing
+ * of dimensions. BatchLinOp allows for storing non-equal objects in the
+ * batches and hence stores a batch_dim object instead of a dim object. The
+ * batch_dim object is optimized to store less amount of data when storing
+ * uniform batches.
+ *
+ * All size validation functions again verify first that the number of batches
+ * are conformant and that the dimensions in the corresponding batches
+ * themselves are also valid/conformant. Here too, optimizations for uniform
+ * batches have been added.
+ *
+ * @ref BatchLinOp
+ */
+class BatchLinOp : public EnableAbstractPolymorphicObject<BatchLinOp> {
+public:
+    /**
+     * Applies a batch linear operator to a batch vector (or a sequence of batch
+     * of vectors).
+     *
+     * Performs the operation x = op(b), where op is this batch linear operator.
+     *
+     * @param b  the input batch vector(s) on which the batch operator is
+     *           applied
+     * @param x  the output batch vector(s) where the result is stored
+     *
+     * @return this
+     */
+    BatchLinOp* apply(ptr_param<const BatchLinOp> b, ptr_param<BatchLinOp> x)
```

Comment:
I don't get why we have a non-const apply. First, I think it makes more sense to only have a const version. And second, we don't have a non-const apply_impl, so even if a linop implementation would want to mutate some internal data during the apply, it couldn't.